### PR TITLE
Use pointer `is_aligned*` methods

### PR DESCRIPTION
**Pull request by WaffleLapkin**
This PR replaces some manual alignment checks with calls to `pointer::{is_aligned, is_aligned_to}` and removes a useless pointer cast.

r? @scottmcm

_split off from #100746_

### CODE REVIEW

File: library/alloc/src/vec/into_iter.rs [hunk 1]
```diff
@@ -148,7 +148,7 @@ impl<T, A: Allocator> Iterator for IntoIter<T, A> {
 
     #[inline]
     fn next(&mut self) -> Option<T> {
-        if self.ptr as *const _ == self.end {
+        if self.ptr == self.end {
```
Comment (WaffleLapkin) [inline library/alloc/src/vec/into_iter.rs:151]: I'd guess this was left from times when `self.ptr` and `self.end` had different types?... But they currently use the same type, so this is useless: https://github.com/rust-lang/rust/blob/de9da0ba8f6627e3aef4c01a95587041867e4625/library/alloc/src/vec/into_iter.rs#L41-L42

**Comment by scottmcm** [comment]
Thanks for splitting these out!

@bors r+ rollup

**Comment by matthiaskrgr** [comment]
@bors r- failed in rollup
https://github.com/rust-lang/rust/pull/100837#issuecomment-1221514220

**Comment by WaffleLapkin** [comment]
I've added the missing feature flag 🥲

**Pull request merged**